src/static/js/caretPosition.js

```diff
@@ -0,0 +1,236 @@
+
+  if(isSelectionInsideTheEditor){
+    // when we have the caret in an empty line, e.g. a line with only a <br>,
+    // getBoundingClientRect() returns all dimensions value as 0
+    var selectionIsInTheBeginningOfLine = range.endOffset > 0;
+    if (selectionIsInTheBeginningOfLine) {
+      var clonedRange = range.cloneRange();
+
+      // we set the selection start and end to avoid error when user selects a text bigger than
+      // the viewport height and uses the arrow keys to expand the selection. In this particular
+      // case is necessary to know where the selections ends because both edges of the selection
+      // is out of the viewport but we only use the end of it to calculate if it needs to scroll
+      clonedRange.setStart(range.endContainer, range.endOffset);
+      clonedRange.setEnd(range.endContainer, range.endOffset);
+
+      rect = clonedRange.getBoundingClientRect();
+      line = {
+        bottom: rect.bottom,
+        height: rect.height,
+        top: rect.top
+      }
+      clonedRange.detach();
+    }
+    // in this case, we can't get the dimension of the element where the caret is
+    if(!rect || rect.height === 0){
+      clonedRange = range.cloneRange();
+      clonedRange.setStart(range.endContainer, range.endOffset);
+      clonedRange.setEnd(range.endContainer, range.endOffset);
+
+      // as we can't get the element height, we create a text node to get the dimensions
+      // on the position
+      var shadowCaret = $(document.createTextNode("|"));
+      clonedRange.insertNode(shadowCaret[0]);
+      clonedRange.selectNode(shadowCaret[0]);
+
+      rect = clonedRange.getBoundingClientRect();
+      line = {
+        bottom: rect.bottom,
+        height: rect.height,
+        top: rect.top
+      }
+      shadowCaret.remove();
+      clonedRange.detach();
+    }
+  }
+  return line;
+}
+
+exports.getPositionOfRepLineAtOffset = function (node, offset) {
+  // it is not a text node, so we cannot make a selection
+  if (node.tagName === 'BR' || node.tagName === 'EMPTY') {
+    return getPositionOfElementOrSelection(node);
+  }
+
+  if (node.length === 0){
+    while (node.length === 0 && node.nextSibling) {
+      node = node.nextSibling;
+    }
+  }
+
+  var newRange = new Range();
+  newRange.setStart(node, offset);
+  newRange.setEnd(node, offset);
+  var linePosition = getPositionOfElementOrSelection(newRange);
+  newRange.detach(); // performance sake
+  return linePosition;
+}
+
+function getPositionOfElementOrSelection(element) {
+  var rect = element.getBoundingClientRect();
+  var linePosition = {
+    bottom: rect.bottom,
+    height: rect.height,
+    top: rect.top
+  }
+  return linePosition;
+}
+
+// here we have two possibilities:
+// [1] the line before the caret line has the same type, so both of them has the same margin, padding
+// height, etc. So, we can use the caret line to make calculation necessary to know where is the top
+// of the previous line
```

Comment:
you are assuming that only line attributes change CSS styles in a more drastic way (margins, paddings, height, etc). This is not necessarily true:

![image](https://user-images.githubusercontent.com/836386/33761334-c8584882-dbef-11e7-9269-57e6d3a59ece.png)

Is your code handling this scenario?

Reply:
It does. We consider the height of the current line to validate if it has space to another line of the same height. When it changes the height of a part (e.g. a word, as the example) it changes the height of the all line.

Reply:
awesome! 👍